vault/plugin_catalog.go

```diff
@@ -35,21 +39,62 @@ type PluginCatalog struct {
 	builtinRegistry BuiltinRegistry
 	catalogView     *BarrierView
 	directory       string
+	logger          log.Logger
+
+	// externalPlugins holds plugin process connections by plugin name
+	//
+	// This allows plugins that suppport multiplexing to use a single grpc
+	// connection to communicate with multiple "backends". Each backend
+	// configuration using the same plugin will be routed to the existing
+	// plugin process.
+	externalPlugins map[string]*externalPlugin
+	mlockPlugins    bool
 
 	lock sync.RWMutex
 }
 
+// pluginClient represents a connection to a plugin process
+type pluginClient struct {
+	logger log.Logger
+
+	// id is the connection ID
+	id string
+
+	// client handles the lifecycle of a plugin process
+	// multiplexed plugins share the same client
+	client      *plugin.Client
+	clientConn  grpc.ClientConnInterface
+	cleanupFunc func() error
+
+	plugin.ClientProtocol
+}
+
+// externalPlugin holds client connections for multiplexed and
+// non-multiplexed plugin processes
+type externalPlugin struct {
```

Comment:
Nit: Would it make sense to move this declaration to be above/before `pluginClient` since `PluginCatalog` hold a map of `externalPlugin` which in turn holds a map of `pluginClient`?